redis/src/main/scala/akka/streams/alpakka/redis/scaladsl/RedisFlow.scala

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright (C) 2016-2018 Lightbend Inc. <http://www.lightbend.com>
+ */
+
+package akka.streams.alpakka.redis.scaladsl
+
+import akka.NotUsed
+import akka.stream.scaladsl.Flow
+import akka.streams.alpakka.redis
+import akka.streams.alpakka.redis.{RedisKeyValue, RedisKeyValues, RedisOperationResult, RedisPubSub}
+import io.lettuce.core.api.StatefulRedisConnection
+import io.lettuce.core.pubsub.StatefulRedisPubSubConnection
+
+import scala.collection.JavaConverters._
+import scala.compat.java8.FutureConverters._
+import scala.concurrent.ExecutionContext
+import scala.util.{Failure, Success}
+
+object RedisFlow {
+
+  def set[K, V](parallelism: Int, connection: StatefulRedisConnection[K, V])(
+      implicit executionContext: ExecutionContext
+  ): Flow[RedisKeyValue[K, V], RedisOperationResult[RedisKeyValue[K, V], String], NotUsed] =
+    Flow[RedisKeyValue[K, V]].mapAsync(parallelism) { redisSet =>
+      val resultFuture = connection.async().set(redisSet.key, redisSet.value).toScala
+      resultFuture.map(r => RedisOperationResult(redisSet, Success(r))).recover {
+        case ex: Throwable => RedisOperationResult(redisSet, Failure(ex))
+      }
+    }
+
+  def mset[K, V](parallelism: Int, connection: StatefulRedisConnection[K, V])(
+      implicit executionContext: ExecutionContext
+  ): Flow[Seq[RedisKeyValue[K, V]], RedisOperationResult[Seq[RedisKeyValue[K, V]], String], NotUsed] =
+    Flow[Seq[RedisKeyValue[K, V]]].mapAsync(parallelism) { sets =>
+      val map = sets.map(f => (f.key, f.value)).toMap.asJava
```

Comment:
1. Can you please elaborate on "directly foreaching over the Seq"? foreach returns `Unit` and why this is better than `asJava` method.  
 
> We could have a flow in Impl that works with the Akka Scala API, but transports lettuce's expected types as stream items

2. This is exactly what I'm doing I have `scalaDsl.RedisFlow` as base implementation and all other Sinks and javadsl flow based on it. 

3. Not all Redis commands are good for items streaming,  mset or hmset works on bulks. It supposed to be bounded array of key values. For example I received response of key /values from one of previous stages. I transfer this collection of key values as single object. Therefore, the inpit for Stage supposed to be collection of items instead of stream of key/values.

Reply:
1. Sure:
```
val map = new java.util.HashMap[K, V]()
sets.foreach(f => map.put(f.key, f.value))
```
This would be to avoid the construction of an intermediate Scala map.

2. Sorry for being vague on my initial comment here. Yes, there is a base implementation in the `scaladsl` package that is used by the `javadsl` implementation and that is great. Looking at the implementation of e.g. `javadsl.RedisFlow.mset` as a whole, it converts the Java List to a Scala Seq first (https://github.com/akka/alpakka/pull/1350/files#diff-2bcc3be581d64d32cd8662ecab9163a8R49), and then converts the Seq to a Java Map (via a Scala Map). To me, it looks like the amount conversions can be reduced: I would suggest to have an internal mset Flow implementation that takes Java Maps, and use that in the public javadsl Flow (converting from a Java List to a Java Map), and also in the scaladsl Flow (converting from a Scala Seq to a Java Map). That way, the amount of collection transformation in the flow could be reduced.

3. mset and hmset are accept streams of item batches, sure. I also saw in the tests that the data type of the mset operation align with the output type of `grouped`, which is great.

Reply:
1. Frankly. I would be a little surprised if I saw such a collection creation using for-each implementation in any production code. 

> val map = new java.util.HashMap[K, V]()
sets.foreach(f => map.put(f.key, f.value))


Assuming that this code will run on Java 10 using G1 GC, I'm not sure that intermediate collection creation is such a big issue. It will be generated in JVM Eden space and will not survive first minor garbage collection. This is the the price that we are paying for working with immutable collections and datatypes. 
We have some production code which performs text analysis on documents using scala immutable collection and with right GC tuning I see 4-5 GigaBytes of String collections disappear from Eden immediately without causing any performance issue on the system. 

I'm really not sure that it worth to change implementation for skipping one immutable collection creation. @ennru  what do you think?